File: ADC-TLA2518.ts
namespace JoyPiAdvanced {

  const ADCPIN = DigitalPin.P16;
  const SYSTEM_STATUS = 0x00;
  const READ_CMD = 0x10;
  const WRITE_CMD = 0x8;
  const CHANNEL_SEL = 0x11;
  const PIN_CFG = 0x5;
  const SEQUENCE_CFG = 0x10;
  const DATA_CFG = 0x2;
  
  function setup(): void {
      pins.spiFrequency(100000);
      pins.spiFormat(8, 0);
      let data = readRegister(SEQUENCE_CFG);
      writeRegister(SEQUENCE_CFG, data & 0xFC);
      writeRegister(PIN_CFG, 0x00);
      writeRegister(SEQUENCE_CFG, 0x00);
  }
  
  function readRegister(reg: number): number {
      let writeBuffer = pins.createBuffer(3);
      let receiveBuffer = pins.createBuffer(1);
      writeBuffer[0] = READ_CMD;
      writeBuffer[1] = reg;
      writeBuffer[2] = 0x00;
      
      pins.digitalWritePin(ADCPIN, 0);
      pins.spiTransfer(writeBuffer, null);
  
      pins.digitalWritePin(ADCPIN, 1);
      control.waitMicros(40);
      pins.digitalWritePin(ADCPIN, 0);
  
      let returnValue = pins.spiWrite(0x00);
      pins.digitalWritePin(ADCPIN, 1);
      control.waitMicros(40);
      return returnValue;
  }
  
  function writeRegister(reg: number, data: number) {
      let writeBuffer = pins.createBuffer(3);
      writeBuffer[0] = WRITE_CMD;
      writeBuffer[1] = reg;
      writeBuffer[2] = data;
  
      pins.digitalWritePin(ADCPIN, 0);
      pins.spiTransfer(writeBuffer, null);
      pins.digitalWritePin(ADCPIN, 1);
      control.waitMicros(40);
  }
  
  function read(channel: number): number {
      // Select channel
      writeRegister(CHANNEL_SEL, channel);
  
      // write empty bytes
      let writeBuffer = pins.createBuffer(2);
      writeBuffer[0] = 0x00;
      writeBuffer[1] = 0x00;
      pins.digitalWritePin(ADCPIN, 0);
      pins.spiTransfer(writeBuffer, null);
      pins.digitalWritePin(ADCPIN, 1);
  
      // read by writing 2 empty bytes
      pins.digitalWritePin(ADCPIN, 0);
      let dataOne = pins.spiWrite(0x00);
      let dataTwo = pins.spiWrite(0x00);
      pins.digitalWritePin(ADCPIN, 1);
      let response = (((dataOne << 8) | dataTwo) >> 4);
      return response;
  }
    
  
    /**
     * Reads the value of the analog digital converter on a specified channel
     * @param channel specifies the channel which is supposed to be read
     */
    //% block="ADC value on channel %channel"
    //% subcategory="Analog-Digital Converter"
    //% channel.min=0 channel.max=7
    //% weight=100
    export function adcReadValue(channel: number): number {
      setup()
      return read(channel)
    }
  
    /**
     * Reads a channel on the analog digital converter and converts the value to a voltage number
     * @param channel specifies the channel which is supposed to be read
     */
    //% block="ADC voltage on channel %channel"
    //% subcategory="Analog-Digital Converter"
    //% channel.min=0 channel.max=7
    //% weight=90
    export function adcReadVoltage(channel: number): number {
      let tempValue = 0
      while(tempValue == 0){
        tempValue = adcReadValue(channel)
      }

      return (tempValue * 5.00/4096.0)
    }
  
  }
  

File: ADC-TLA2518.cpp
#include "pxt.h"
#include <SPI.h>

#define ADCPIN P16
#define SYSTEM_STATUS 0x00
#define READ_CMD 0x10
#define WRITE_CMD 0x08
#define CHANNEL_SEL 0x11
#define PIN_CFG 0x05
#define SEQUENCE_CFG 0x10

using namespace pxt;
namespace TLA2518 {

    MicroBitPin *adcPin = &uBit.io.P16;
    SPI spi(ADCPIN);

    uint8_t readRegister(uint8_t reg) {
        adcPin->setDigitalValue(0);
        spi.write(READ_CMD);
        spi.write(reg);
        spi.write(0x00);
        uBit.sleep_us(200);
        adcPin->setDigitalValue(1);
        adcPin->setDigitalValue(0);
        uint8_t returnValue = spi.write(0x00);
        adcPin->setDigitalValue(1);
        uBit.sleep_us(100);
        return returnValue;
    }

    void writeRegister(uint8_t reg, uint8_t data) {
        adcPin->setDigitalValue(0);
        spi.write(WRITE_CMD);
        spi.write(reg);
        spi.write(data);
        uBit.sleep_us(100);
        adcPin->setDigitalValue(1);
    }

    uint8_t read(uint8_t channel) {
        uint8_t data = readRegister(SEQUENCE_CFG);
        writeRegister(SEQUENCE_CFG, data & 0xFC);
        writeRegister(PIN_CFG, 0x00);
        writeRegister(SEQUENCE_CFG, 0x00);
        writeRegister(CHANNEL_SEL, channel);
        adcPin->setDigitalValue(0);
        spi.write(0x00);
        spi.write(0x00);
        adcPin->setDigitalValue(1);
        adcPin->setDigitalValue(0);
        uint8_t dataOne = spi.write(0x00);
        uint8_t dataTwo = spi.write(0x00);
        adcPin->setDigitalValue(1);
        return ((dataOne << 8) | dataTwo) >> 4;
    }

    /*uint8_t adcReadValue(uint8_t channel) {
        return read(channel);
    }*/

    uint8_t ValueZero(){
        return read(0)
    }

    uint8_t ValueOne(){
        return read(1)
    }

    uint8_t ValueTwo(){
        return read(2)
    }

    uint8_t ValueThree(){
        return read(3)
    }

    uint8_t ValueFour(){
        return read(4)
    }

    uint8_t ValueFive(){
        return read(5)
    }

    uint8_t ValueSix(){
        return read(6)
    }

    uint8_t ValueSeven(){
        return read(7)
    }

}